counsel.el

```diff
@@ -2695,6 +2718,12 @@ This uses `counsel-ag' with `counsel-ack-base-command' replacing
 Note: don't use single quotes for the regex."
   :type 'string)
 
+(defcustom counsel-rg-version nil
+  "Ripgrep version used by `counsel-rg'.
+
+Version 0.10.0 and later supports PCRE2."
+  :type 'string)
```

Comment:
User options are not like variables in that they shouldn't be left empty and set later on by the program. First of all, this breaks the `:type` contract and isn't documented in the docstring in this case. More importantly, programs should strive to never `let`-bind user options, let alone change their value altogether - they are declarative contracts between the user and the program.

Closer to what you seem to want to do is to give `counsel-rg-version` a special value which says "figure out the version automatically". This could either be `nil` or, say, a symbol named `auto`. Then the user option would look like this:

```el
(defcustom counsel-rg-version 'auto
  "Ripgrep version used by `counsel-rg' as a string.
The symbol `auto' means determine the version automatically.
Version 0.10.0 and later supports PCRE2."
  :type '(choice string (const auto)))
```

But I still don't understand what the benefit of providing this user option is. Why can't we just act as if it's `auto` all the time? Users can still disable this matching by setting `counsel-ag-look-around` to `nil`.

Asking users to specify the version of a program seems like a very low-level implementation detail. If we really need to provide a user option, why can't we just provide a toggle for whether or not to use PCRE2 where possible?

Reply:
> User options are not like variables in that they shouldn't be left empty and set later on by the program.

point taken.

> But I still don't understand what the benefit of providing this user option is. Why can't we just act as if it's `auto` all the time?

Because there are 100's of scenarios where `auto` will not work. What if the user renames `rg` to `ripgrep`. What if the version they want to use is not in their `$PATH`? what if they built their own executable that changes how the version is displayed? Automatically setting it will work 99% of the time, but the user needs an option when it doesn't work. The worst thing we can do is have someone use `counsel-rg` and it doesn't work.

> Users can still disable this matching by setting `counsel-ag-look-around` to `nil`.

That will not work. `counsel-ag-look-around` is **not** a user option. Neither is `counsel-ag-command`. They are let bound to control the behavior `counsel-ag`. Though I agree that it might be better to create a subroutine rather then wrapping `counsel-ag`, That is the not design pattern already in place. 

The idea behind this implementation is that the user should not care about enabling look-arounds. If the user wants to disable look-arounds in line search tools, then they already have a mechanism in place; `(add-to-list 'ivy-re-builder-alist '(counsel-ag . ivy--regex))`. This will use a regex builder that does not require look-arounds. This is why I wanted to change remove the `:caller`, so that you could bind each of the line search tools to their own regex builder. 

The implementation details about when to use look-arounds should be hidden from the user. If they are using a multi-pass regex builder like `ivy--regex-plus` or `ivy--regex-ignore-order` then it will "just work", if they are using a single pass regex builder, it will "just work".

> Asking users to specify the version of a program seems like a very low-level implementation detail. If we really need to provide a user option, why can't we just provide a toggle for whether or not to use PCRE2 where possible?

Version is a low level detail. But even if we used an option instead to control PCRE enablement, we still need to know if the version of `rg` has PCRE2 support. If the user says "use PCRE2 where possible", that tells us nothing, because we still don't know if it is possible to use in `rg` until we know the version.

 So instead of just asking "what version of rg are you using?" (which the user won't need to care about 99% of the time) We would have to ask "does your version of rg support PCRE2?". I feel like the first question is much easier for someone to answer then the second, which requires digging into the release notes.

Based on the feedback given, I am going to remove the user option `counsel-rg-version` and instead just bind it to a `defvar`. We use a robust `auto` solution to set it. On the off chance that user is:
- using `counsel-rg`
- using a version greater then 0.10.0
- has a rg version was not auto detected
- is using a multi-pass regex builder

Then they will have to look into the issue and set `counsel-rg-version`. However I think this senario will be quite rare.

Thank you for your feedback.

Reply:
> Because there are 100's of scenarios where `auto` will not work. What if the user renames `rg` to `ripgrep`.

The standard solution to this is to provide a user option called `counsel-rg-program` or similar. The information can already be extracted from the existing user option `counsel-rg-base-command`. Program names shouldn't be hard-coded.

> What if the version they want to use is not in their `$PATH`?

Then `counsel-rg` is anyway not going to work until the user fixes their setup.

> what if they built their own executable that changes how the version is displayed?

Then we can't assume it supports any other `ripgrep` options anyway and have to fall back to some other sensible behaviour, such as assuming `--pcre2` isn't supported or similar. The boolean user option or variable I described previously could be used to override automatic detection. This would let the user emphatically say "my program supports / doesn't support --pcre2, I know what I'm doing".

> Automatically setting it will work 99% of the time, but the user needs an option when it doesn't work. The worst thing we can do is have someone use `counsel-rg` and it doesn't work.

I'm not against providing a user option. I just don't see how asking the user for an arbitrary program version is any better than providing a simpler user option which determines whether to go ahead with our assumptions or not. One is an implementation detail, the other is a high-level declaration.

> This is why I wanted to change remove the `:caller`, so that you could bind each of the line search tools to their own regex builder.

I still think this is a bad idea. I consider the inability to customise the regexp builder of `counsel-rg` a bug with the implementation of `counsel-rg`, not a shortcoming of `counsel-ag`. Users of `counsel-ag` do not deserve to be punished for the way `counsel-rg` was originally implemented.

> The implementation details about when to use look-arounds should be hidden from the user. As the title says, "make counsel-ag respect ivy-re-builder-alist". If they are using a multi-pass regex builder like `ivy--regex-plus` or `ivy--regex-ignore-order` then it will "just work", if they are using a single pass regex builder, it will "just work".

Fair enough. I wasn't suggest `counsel-ag-look-around` was a panacea; I was just giving an example using what was already there. 

> Version is a low level detail. But even if we used an option instead to control PCRE enablement, we still need to know if the version of `rg` has PCRE2 support. If the user says "use PCRE2 where possible", that tells us nothing, because we still don't know if it is possible to use in `rg` until we know the version.

It does tell us something though - it tells us we can assume PCRE2 is supported until proven otherwise, i.e. until the user disables this feature. The docstring would explain as much. Though a better user option would be a list of switches (e.g. `'("--pcre2")`) that the user could customise to their heart's content.

> So instead of just asking "what version of rg are you using?" (which the user won't need to care about 99% of the time) We would have to ask "does your version of rg support PCRE2?". I feel like the first question is much easier for someone to answer then the second, which requires digging into the release notes.

Not really - it would be easy to tell because the program would exit with an error code. Though now that I think about it more, a list of switches would be better than asking "does your version support PCRE2", as a list handles all possible present and future scenarios of switch customisation. The docstring could then explain troubleshooting caveats such as the lack of PCRE2 support in older versions.

Reply:
Thank you for taking time to reply. 

I pushed a new commit that should address your feedback.